awscli/customizations/history/show.py

```diff
@@ -0,0 +1,454 @@
+                    request_id] = self._num_api_calls
+                self._num_api_calls += 1
+            return self._request_id_to_api_num[request_id]
+
+    def _format_value_with_description(self, value, value_description,
+                                       event_record, value_format=None,
+                                       description_color=_DESCRIPTION_COLOR):
+        formatted_value_with_description = self._color_description(
+            value_description + ': ', description_color)
+        formatted_value = self._format_value(value, event_record, value_format)
+        formatted_value_with_description += self._color_description_value(
+            formatted_value)
+        formatted_value_with_description += '\n'
+        return formatted_value_with_description
+
+    def _format_value(self, value, event_record, value_format=None):
+        if value_format:
+            return getattr(self, '_format_' + value_format)(
+                value, event_record)
+        return str(value)
+
+    def _format_timestamp(self, event_timestamp, event_record=None):
+        return datetime.datetime.fromtimestamp(
+            event_timestamp/1000.0).strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]
+
+    def _format_dictionary(self, obj, event_record=None):
+        return json.dumps(obj=obj, sort_keys=True, indent=4)
+
+    def _format_http_body(self, body, event_record):
+        if not body:
+            return 'There is no associated body'
+        elif event_record['payload'].get('streaming', False):
+            return 'The body is a stream and will not be displayed'
+        elif self._is_xml(body):
+            return self._format_http_body_xml(body)
+        elif self._is_json_structure(body):
+            return self._format_http_body_json(body)
+        else:
+            return body
+
+    def _format_http_body_xml(self, body):
+        stripped_body = self._strip_whitespace(body)
+        xml_dom = xml.dom.minidom.parseString(stripped_body)
+        return xml_dom.toprettyxml(indent=' '*4, newl='\n')
+
+    def _format_http_body_json(self, body):
+        obj = json.loads(body)
+        return self._format_dictionary(obj)
+
+    def _color_title(self, text):
+        return self._color_if_configured(text, self._TITLE_COLOR)
+
+    def _color_description(self, text, color=_DESCRIPTION_COLOR):
+        return self._color_if_configured(text, color)
+
+    def _color_description_value(self, text):
+        return self._color_if_configured(text, self._DESCRIPTION_VALUE_COLOR)
+
+    def _color_if_configured(self, text, color):
+        if self._colorize:
+            return color + text + colorama.Style.RESET_ALL
+        return text
+
+    def _is_xml(self, body):
+        try:
+            xml.dom.minidom.parseString(body)
+        except xml.parsers.expat.ExpatError:
+            return False
+        return True
+
+    def _strip_whitespace(self, xml_string):
+        xml_dom = xml.dom.minidom.parseString(xml_string)
```

Comment:
I will add a comment here. But the main reason was because services like ec2 will return xml that is already pretty and as a result blindly making it pretty again adds even more whitespace and newlines on top of what was already there.